crates/oxc_linter/src/rules/typescript/prefer_function_type.rs

```diff
@@ -0,0 +1,743 @@
+use oxc_ast::{
+    ast::{ExportDefaultDeclarationKind, Expression, TSInterfaceDeclaration, TSSignature, TSType},
+    AstKind, CommentKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, fixer::Fix, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("typescript-eslint(prefer-function-type): Enforce using function types instead of interfaces with call signatures.")]
+#[diagnostic(
+    severity(warning),
+    help("See https://typescript-eslint.io/rules/prefer-function-type for documentation.")
+)]
+struct PreferFunctionTypeDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct PreferFunctionType;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Enforce using function types instead of interfaces with call signatures.
+    /// TypeScript allows for two common ways to declare a type for a function:
+    ///
+    /// - Function type: `() => string`
+    /// - Object type with a signature: `{ (): string }`
+    ///
+    /// The function type form is generally preferred when possible for being more succinct.
+    ///
+    /// This rule suggests using a function type instead of an interface or object type literal with a single call signature.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // error
+    /// interface Example {
+    ///   (): string;
+    /// }
+    ///
+    /// function foo(example: { (): number }): number {
+    ///   return example();
+    /// }
+    ///
+    /// interface ReturnsSelf {
+    ///   (arg: string): this;
+    /// }
+    ///
+    /// // success
+    /// type Example = () => string;
+    ///
+    /// function foo(example: () => number): number {
+    ///   return bar();
+    /// }
+    ///
+    /// // returns the function itself, not the `this` argument.
+    /// type ReturnsSelf = (arg: string) => ReturnsSelf;
+    ///
+    /// function foo(bar: { (): string; baz: number }): string {
+    ///   return bar();
+    /// }
+    ///
+    /// interface Foo {
+    ///   bar: string;
+    /// }
+    /// interface Bar extends Foo {
+    ///   (): void;
+    /// }
+    ///
+    /// // multiple call signatures (overloads) is allowed:
+    /// interface Overloaded {
+    ///   (data: string): number;
+    ///   (id: number): string;
+    /// }
+    /// // this is equivalent to Overloaded interface.
+    /// type Intersection = ((data: string) => number) & ((id: number) => string);
+    /// ```
+    PreferFunctionType,
+    correctness
```

Comment:
According to https://typescript-eslint.io/rules/prefer-function-type, this is a `style` rule.